File: app/src/main/java/com/close/hook/ads/hook/util/ContextUtil.java
package com.close.hook.ads.hook.util;

import android.app.Application;
import android.content.Context;
import android.content.ContextWrapper;
import android.app.Instrumentation;
import java.util.concurrent.ConcurrentLinkedQueue;
import java.util.concurrent.atomic.AtomicBoolean;
import de.robv.android.xposed.XposedBridge;

public class ContextUtil {

    public static volatile Context activityThreadContext;
    public static volatile Context applicationContext;
    public static volatile Context contextWrapperContext;
    public static volatile Context instrumentationContext;

    private static final ConcurrentLinkedQueue<Runnable> activityThreadContextCallbacks = new ConcurrentLinkedQueue<>();
    private static final ConcurrentLinkedQueue<Runnable> applicationContextCallbacks = new ConcurrentLinkedQueue<>();
    private static final ConcurrentLinkedQueue<Runnable> contextWrapperContextCallbacks = new ConcurrentLinkedQueue<>();
    private static final ConcurrentLinkedQueue<Runnable> instrumentationContextCallbacks = new ConcurrentLinkedQueue<>();

    private static final AtomicBoolean isActivityThreadContextInitialized = new AtomicBoolean(false);
    private static final AtomicBoolean isApplicationContextInitialized = new AtomicBoolean(false);
    private static final AtomicBoolean isContextWrapperContextInitialized = new AtomicBoolean(false);
    private static final AtomicBoolean isInstrumentationContextInitialized = new AtomicBoolean(false);

    private static final String TAG = "ContextUtil";

    public static void initialize(Runnable initialCallback) {
        if (initialCallback != null) {
            addOnActivityThreadContextInitializedCallback(initialCallback);
            addOnApplicationContextInitializedCallback(initialCallback);
            addOnContextWrapperContextInitializedCallback(initialCallback);
            addOnInstrumentationContextInitializedCallback(initialCallback);
        }
        initializeContextHooks();
    }

    private static void initializeContextHooks() {
        HookUtil.hookAllMethods(
            "android.app.ActivityThread",
            "performLaunchActivity",
            "after",
            param -> {
                Object activity = param.getResult();
                if (activity instanceof Context) {
                    activityThreadContext = (Context) activity;
                    triggerCallbacksIfInitialized(activityThreadContext, isActivityThreadContextInitialized, activityThreadContextCallbacks, "ActivityThreadContext");
                }
            }
        );

        HookUtil.findAndHookMethod(
            Application.class,
            "attach",
            new Object[]{Context.class},
            "after",
            param -> {
                applicationContext = (Context) param.args[0];
                triggerCallbacksIfInitialized(applicationContext, isApplicationContextInitialized, applicationContextCallbacks, "ApplicationContext");
            }
        );

        HookUtil.findAndHookMethod(
            ContextWrapper.class,
            "attachBaseContext",
            new Object[]{Context.class},
            "after",
            param -> {
                contextWrapperContext = (Context) param.args[0];
                triggerCallbacksIfInitialized(contextWrapperContext, isContextWrapperContextInitialized, contextWrapperContextCallbacks, "ContextWrapperContext");
            }
        );

        HookUtil.findAndHookMethod(
            Instrumentation.class,
            "callApplicationOnCreate",
            new Object[]{Application.class},
            "after",
            param -> {
                instrumentationContext = (Application) param.args[0];
                triggerCallbacksIfInitialized(instrumentationContext, isInstrumentationContextInitialized, instrumentationContextCallbacks, "InstrumentationContext");
            }
        );
    }

    private static void triggerCallbacksIfInitialized(Context context, AtomicBoolean isInitialized, ConcurrentLinkedQueue<Runnable> callbacks, String contextType) {
        if (context != null && isInitialized.compareAndSet(false, true)) {
            XposedBridge.log(TAG + " | Initialized (" + contextType + "): " + context);
            Runnable callback;
            while ((callback = callbacks.poll()) != null) {
                callback.run();
            }
        }
    }

    public static void addOnActivityThreadContextInitializedCallback(Runnable callback) {
        addCallback(callback, isActivityThreadContextInitialized, activityThreadContextCallbacks);
    }

    public static void addOnApplicationContextInitializedCallback(Runnable callback) {
        addCallback(callback, isApplicationContextInitialized, applicationContextCallbacks);
    }

    public static void addOnContextWrapperContextInitializedCallback(Runnable callback) {
        addCallback(callback, isContextWrapperContextInitialized, contextWrapperContextCallbacks);
    }

    public static void addOnInstrumentationContextInitializedCallback(Runnable callback) {
        addCallback(callback, isInstrumentationContextInitialized, instrumentationContextCallbacks);
    }

    private static void addCallback(Runnable callback, AtomicBoolean isInitialized, ConcurrentLinkedQueue<Runnable> callbacks) {
        if (isInitialized.get()) {
            callback.run();
        } else {
            callbacks.offer(callback);
        }
    }
}

File: app/src/main/java/com/close/hook/ads/hook/util/ContextUtil.kt
package com.close.hook.ads.hook.util

import android.app.Application
import android.app.Instrumentation
import android.content.Context
import android.content.ContextWrapper
import de.robv.android.xposed.XposedBridge
import java.util.concurrent.ConcurrentLinkedQueue
import java.util.concurrent.atomic.AtomicBoolean

object ContextUtil {

    private const val TAG = "ContextUtil"

    private class ContextState {
        val isInitialized = AtomicBoolean(false)
        val callbacks = ConcurrentLinkedQueue<() -> Unit>()

        fun triggerCallbacks(context: Context, contextType: String) {
            if (isInitialized.compareAndSet(false, true)) {
                XposedBridge.log("$TAG | Context Initialized ($contextType): $context")
                callbacks.forEach { it() }
                callbacks.clear()
            }
        }

        fun addCallback(callback: () -> Unit, currentContext: Context?) {
            if (isInitialized.get() && currentContext != null) {
                callback()
            } else {
                callbacks.offer(callback)
            }
        }
    }

    private val activityThreadState = ContextState()
    private val applicationState = ContextState()
    private val contextWrapperState = ContextState()
    private val instrumentationState = ContextState()

    @JvmField
    @Volatile
    var activityThreadContext: Context? = null

    @JvmField
    @Volatile
    var applicationContext: Context? = null

    @JvmField
    @Volatile
    var contextWrapperContext: Context? = null

    @JvmField
    @Volatile
    var instrumentationContext: Context? = null

    fun setupContextHooks() {
        HookUtil.hookAllMethods(
            "android.app.ActivityThread",
            "performLaunchActivity",
            "after"
        ) { param ->
            (param.result as? Context)?.let { context ->
                activityThreadContext = context
                activityThreadState.triggerCallbacks(context, "ActivityThreadContext")
            }
        }

        HookUtil.findAndHookMethod(
            Application::class.java,
            "attach",
            arrayOf(Context::class.java),
            "after"
        ) { param ->
            (param.args[0] as? Context)?.let { context ->
                applicationContext = context
                applicationState.triggerCallbacks(context, "ApplicationContext")
            }
        }

        HookUtil.findAndHookMethod(
            ContextWrapper::class.java,
            "attachBaseContext",
            arrayOf(Context::class.java),
            "after"
        ) { param ->
            (param.args[0] as? Context)?.let { context ->
                contextWrapperContext = context
                contextWrapperState.triggerCallbacks(context, "ContextWrapperContext")
            }
        }

        HookUtil.findAndHookMethod(
            Instrumentation::class.java,
            "callApplicationOnCreate",
            arrayOf(Application::class.java),
            "after"
        ) { param ->
            (param.args[0] as? Application)?.let { context ->
                instrumentationContext = context
                instrumentationState.triggerCallbacks(context, "InstrumentationContext")
            }
        }
    }

    fun addOnActivityThreadContextInitializedCallback(callback: () -> Unit) {
        activityThreadState.addCallback(callback, activityThreadContext)
    }

    fun addOnApplicationContextInitializedCallback(callback: () -> Unit) {
        applicationState.addCallback(callback, applicationContext)
    }

    fun addOnContextWrapperContextInitializedCallback(callback: () -> Unit) {
        contextWrapperState.addCallback(callback, contextWrapperContext)
    }

    fun addOnInstrumentationContextInitializedCallback(callback: () -> Unit) {
        instrumentationState.addCallback(callback, instrumentationContext)
    }
}
